### PR TITLE
TandemFoil: AdamW LR sweep with slices=64 (3e-4, 5e-4, 8e-4, 1e-3)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

AdamW at lr=5e-4 demolished Lion on AirfRANS (-38%). Round 1 established an AdamW baseline on TandemFoil at only 2 epochs (val surface_pressure_mae=269.32) due to the high per-epoch cost (~15 min/epoch at default slices=96). This PR tests whether AdamW also dominates on TandemFoil, and finds the optimal LR, using slices=64 to fit more epochs within the timeout.

**Key question:** Does reducing slices 96→64 on TandemFoil allow enough extra epochs to show a meaningful LR sensitivity difference?

## Instructions

Run 4 parallel trials on 4 GPUs. All use `model_slices=64` (faster epochs), no physics features, surface_refine=True.

```bash
cd target/icml2026

# v0: AdamW lr=3e-4
CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset tandemfoil \
  --optimizer adamw --lr 3e-4 --model_slices 64 \
  --wandb_group tandem-adamw-lr-sweep \
  --wandb_run_name v0-adamw-lr3e4 &

# v1: AdamW lr=5e-4 (same as baseline but with slices=64)
CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset tandemfoil \
  --optimizer adamw --lr 5e-4 --model_slices 64 \
  --wandb_group tandem-adamw-lr-sweep \
  --wandb_run_name v1-adamw-lr5e4 &

# v2: AdamW lr=8e-4
CUDA_VISIBLE_DEVICES=2 python train.py \
  --dataset tandemfoil \
  --optimizer adamw --lr 8e-4 --model_slices 64 \
  --wandb_group tandem-adamw-lr-sweep \
  --wandb_run_name v2-adamw-lr8e4 &

# v3: AdamW lr=1e-3
CUDA_VISIBLE_DEVICES=3 python train.py \
  --dataset tandemfoil \
  --optimizer adamw --lr 1e-3 --model_slices 64 \
  --wandb_group tandem-adamw-lr-sweep \
  --wandb_run_name v3-adamw-lr1e3 &

wait
```

Report: epochs completed, `val_primary/surface_pressure_mae` at each epoch (it's `val_eq4/surface_pressure_mae` in the logs). Also report per-split MAEs: `val_single_in_dist`, `val_geom_camber_rc`, `val_geom_camber_cruise`, `val_re_rand`. Note whether any Infinity values appear in the metrics (seen in baseline at epoch 2, likely an EMA artifact).

## Baseline

- **val_primary/surface_pressure_mae:** 269.32 (val) / 262.56 (test)
- **Config:** AdamW lr=5e-4, slices=96 (default), 2 epochs only
- **PR:** #2416 (alphonse, W&B run: r5t674uy)
- **Reproduce:** `cd target/icml2026 && python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4`

## Results

_To be filled in by student_

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

### Summary

| Run | LR | Epochs | val_eq4/surface_pressure_mae | test_eq4/surface_pressure_mae | W&B run |
|-----|------|--------|------------------------------|-------------------------------|---------|
| v0 | 3e-4 | 1 | 444.4 | 434.1 | `8vx7b9il` |
| v1 | 5e-4 | 1 | 338.2 | 326.3 | `6uhmqf7u` |
| **v2** | **8e-4** | **2** | **254.3** | **249.8** | **`fz7ekben`** |
| v3 | 1e-3 | 2 | 456.9 | 433.0 | `b9fmgm1i` |
| Baseline | 5e-4 | 2 (slices=96) | 269.3 | 262.6 | `r5t674uy` |

### Per-split test MAE (surface_pressure_mae)

| Run | LR | single_in_dist | geom_camber_rc | geom_camber_cruise | re_rand |
|-----|------|----------------|----------------|--------------------|---------|
| v0 | 3e-4 | 468.0 | 346.2 | 474.3 | 448.0 |
| v1 | 5e-4 | 394.4 | 369.0 | 244.5 | 297.2 |
| **v2** | **8e-4** | **266.4** | **230.3** | **246.9** | **255.6** |
| v3 | 1e-3 | 519.3 | 379.2 | 416.4 | 417.1 |
| Baseline | 5e-4 | 304.8 | 270.9 | 224.6 | 249.9 |

### Epoch-by-epoch (v2 and v3 got 2 epochs; v0 and v1 only 1)

| Run | Epoch 1 val MAE | Epoch 2 val MAE | Epoch 1 loss | Epoch 2 loss |
|-----|-----------------|-----------------|--------------|--------------|
| v0 (3e-4) | 444.4 | — | 1.078 | — |
| v1 (5e-4) | 338.2 | — | 1.098 | — |
| v2 (8e-4) | 402.0 | **254.3** | 1.195 | 0.778 |
| v3 (1e-3) | 843.2 | 456.9 | 1.172 | 0.743 |

### Commands used

```bash
cd target/icml2026

# All runs used: --model-slices 64 --surface-refine --optimizer adamw --epochs 50
# SENPAI_TIMEOUT_MINUTES=30, SENPAI_MAX_EPOCHS=50

PYTHONUNBUFFERED=1 CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer adamw --lr 3e-4 --model-slices 64 --epochs 50 --surface-refine --agent alphonse --wandb-name "alphonse/v0-adamw-lr3e4" --wandb-group tandem-adamw-lr-sweep &
PYTHONUNBUFFERED=1 CUDA_VISIBLE_DEVICES=1 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --model-slices 64 --epochs 50 --surface-refine --agent alphonse --wandb-name "alphonse/v1-adamw-lr5e4" --wandb-group tandem-adamw-lr-sweep &
PYTHONUNBUFFERED=1 CUDA_VISIBLE_DEVICES=2 python train.py --dataset tandemfoil --optimizer adamw --lr 8e-4 --model-slices 64 --epochs 50 --surface-refine --agent alphonse --wandb-name "alphonse/v2-adamw-lr8e4" --wandb-group tandem-adamw-lr-sweep &
PYTHONUNBUFFERED=1 CUDA_VISIBLE_DEVICES=3 python train.py --dataset tandemfoil --optimizer adamw --lr 1e-3 --model-slices 64 --epochs 50 --surface-refine --agent alphonse --wandb-name "alphonse/v3-adamw-lr1e3" --wandb-group tandem-adamw-lr-sweep &
wait
```

### Peak memory

Not directly reported, but GPU memory ranged from 57-80 GB across the 4 runs during training.

### Infinity values

All 4 runs show `Infinity` in `test_geom_camber_cruise/mae_vol_p` and `volume_mae` — consistent EMA artifact from the baseline, not LR-specific.

### What happened

**v2 (lr=8e-4) is the clear winner**, beating the baseline on both val (254.3 vs 269.3, -5.6%) and test (249.8 vs 262.6, -4.9%). The key factor: v2 and v3 completed epoch 1 just before the 30-min timeout, allowing them to run epoch 2, while v0 and v1 hit the timeout after epoch 1. The extra epoch was decisive — v2's epoch 1 val MAE (402.0) was worse than v1's (338.2), but epoch 2 brought it down dramatically to 254.3.

The LR ranking at epoch 1 was: 5e-4 > 8e-4 > 3e-4 >> 1e-3. But with 2 epochs, 8e-4 surpassed 5e-4. This suggests 8e-4 benefits more from continued training — the higher learning rate makes larger initial steps that pay off over time. lr=1e-3 was clearly too aggressive (epoch 1 val MAE of 843).

The epoch count difference (1 vs 2) is largely due to I/O contention from 4 parallel runs sharing one disk, which inflated per-epoch time to ~25-30 min (vs ~15 min for a single run at slices=96). v2 and v3 happened to complete their epoch 1 validation slightly faster than the others, slipping under the 30-min timeout.

Note: slices=64 vs baseline's slices=96 means fewer attention slices, so the model has less capacity. Despite this, v2 beat the baseline, suggesting the extra epoch matters more than model capacity at this stage.

### Suggested follow-ups

1. **Run lr=8e-4 with slices=96 (default)** to see if the higher capacity model + optimal LR yields even better results with more epochs.
2. **Run lr=5e-4 and lr=8e-4 sequentially (not parallel)** for more epochs — the I/O contention from parallel runs severely limited epoch count. A single run could likely fit 3-4 epochs in 30 min.
3. **Try lr=6e-4 and lr=7e-4** to narrow the sweet spot between 5e-4 and 8e-4.
4. **Cosine schedule interaction**: the baseline uses cosine_t_max=150. With only 2 epochs, the LR barely decays. A shorter cosine schedule might help converge faster.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2433 because ADVISOR: Closing — best AdamW result (lr=8e-4, val=254.34) does not beat Lion baseline (197.87). **Key finding: Lion dominates AdamW on TandemFoil** (opposite of AirfRANS). AdamW lr=8e-4 is the optimal AdamW LR but still 22% behind Lion. Also confirmed: running 4 parallel jobs caused epoch starvation (v0/v1 only got 1 epoch at slices=64). The AdamW lr=8e-4 signal will be useful if future architecture changes favor AdamW. Thank you alphonse..
